### PR TITLE
drivers: flash: spi: Move to using select in Kconfig for SPI bus

### PR DESCRIPTION
**Pull request by pdgendt**
* Updated to the latest policy to `select` the SPI bus for flash controllers
* Updated SPI flash samples to remove default selected symbols

### CODE REVIEW

File: boards/arm/arty/Kconfig.defconfig
Comment (de-nordic) [inline boards/arm/arty/Kconfig.defconfig:30]: Is removal of this correct? y here depends on another symbol.

Comment (pdgendt) [reply boards/arm/arty/Kconfig.defconfig:30]: It is required to have `SPI_NOR` select `SPI`, otherwise a dependency loop exists:
`SPI_XLNX_AXI_QUADSPI depends on SPI -> SPI_NOR selects SPI -> SPI_NOR depends on SPI_XLNX_AXI_QUADSPI`.

I must admit the change for `SPI_NOR` went a bit over my head, maybe I should only have `SPI_FLASH_AT45` in this PR (which is what I am working with).

**Review by MaureenHelm — DISMISSED** [review]

File: boards/arm/arty/Kconfig.defconfig
Comment (de-nordic) [reply boards/arm/arty/Kconfig.defconfig:30]: I am fine with the SPI change, in flash drivers, but I am not sure how to address board configurations.
Board configurations  may still want to select `SPI_NOR`, to enable NOR flash support by default, when FLASH is enabled.

The `default y if SPI_XLNX_AXI_QUADSPI`  can be false if somebody disables it and `SPI_NOR`, here, may depend on `SPI_XLNX_AXI_QUADSPI` because this might be the SPI driver, the SPI_NOR hangs on.

Anyway, this requires some comments from board maintainers.

**Review by galak — DISMISSED** [review]

File: boards/arm/arty/Kconfig.defconfig
Comment (pdgendt) [reply boards/arm/arty/Kconfig.defconfig:30]: The cases where I removed the board default for `SPI_NOR` were only for `if FLASH ... endif` I think.

**Comment by pdgendt** [comment]
> Marking DNM to allow some board maintainer feedback based on @de-nordic's comment.

@galak Anyone in particular we should ping?

**Comment by pdgendt** [comment]
@galak is the DNM label still applicable?

**Comment by galak** [comment]
@henrikbrixandersen, @dleach02 , @danieldegrasse can you guys take a look

**Comment by galak** [comment]
@pdgendt  looks like this needs a rebase to resolve a merge conflict.

File: drivers/flash/Kconfig.at45
Comment (nordicjm) [inline drivers/flash/Kconfig.at45:10]: Why has this change been made? This should not select SPI, this originally correctly depends on SPI

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: drivers/flash/Kconfig.at45
Comment (nordicjm) [reply drivers/flash/Kconfig.at45:10]: CC @tejlmand

Comment (pdgendt) [reply drivers/flash/Kconfig.at45:10]: The policy is to use `select` instead of `depends on` if a peripheral uses a bus. See [this commit](https://github.com/zephyrproject-rtos/zephyr/commit/df81fef94483da9f2811d48088affbcfd61ab18c) for more info.

Comment (nordicjm) [reply drivers/flash/Kconfig.at45:10]: That specifically states for devices that can be on an SPI and/or I2C, of which the at45 cannot as it is SPI only

File: drivers/flash/Kconfig.nor
Comment (nordicjm) [inline drivers/flash/Kconfig.nor:11]: Likewise with this, as the name implies SPI_NOR is SPI, it cannot reside on an i2c bus

**Review by nordicjm — DISMISSED** [review]

File: drivers/flash/Kconfig.at45
Comment (pdgendt) [reply drivers/flash/Kconfig.at45:10]: Yes, but it has been generally applied to all peripherals because it reduces the need to have additional `kconfig` symbols defined, which can be deduced from the device tree.
As the following search shows (these commits refer to the same commit):
```
$ git log --oneline --grep "df81fef"
d36125654e drivers: ethernet: Update Kconfig
a50814dc9d drivers: w1: use select SERIAL in Kconfig for w1-zephyr-serial driver
baaf904885 video: Move to using select in Kconfig for I2C bus
b37f393079 drivers: audio: Update Kconfig
e438da8a36 adc: Move to using select in Kconfig for I2C & SPI busses
18a1db193d drivers: dac: Update Kconfig for 'select I2C/SPI'
465b7615a9 drivers: w1: Move to using select in Kconfig for I2C bus
9e9aab3bcf drivers: can: Move to using select in Kconfig for SPI bus
a96ef8a631 drivers: misc: ft8xx: Update Kconfig
d9882ff422 eeprom: Move to using select in Kconfig for I2C & SPI busses
616342d58b drivers: ieee802154: Move to using select in Kconfig for SPI bus
db0a8c63fe drivers: led: Update Kconfig
287043cddb drivers: kscan: Update Kconfig
902f42f491 drivers: display: Update Kconfig
```

Comment (pdgendt) [reply drivers/flash/Kconfig.at45:10]: > Why has this change been made? This should not select SPI, this originally correctly depends on SPI

@galak If this is the confirmed, I will close the PR.

Comment (nordicjm) [reply drivers/flash/Kconfig.at45:10]: I will leave it for @tejlmand to comment on this. This is the exact issue that has to be fixed elsewhere because select is the wrong thing to use, depends on is correct e.g. https://github.com/zephyrproject-rtos/zephyr/pull/53778/files

Comment (henrikbrixandersen) [reply drivers/flash/Kconfig.at45:10]: The policy to use `select` over `depends on` for bus dependencies was discussed in https://github.com/zephyrproject-rtos/zephyr/pull/48707. The rationale is, that if a user enabled a Kconfig for a given driver, surely they also want to enable to needed bus driver class.

Comment (tejlmand) [reply drivers/flash/Kconfig.at45:10]: @henrikbrixandersen not sure why I was not included in the discussion of #48707 ?

There are some risks associated with such a transition to `select` compared to a `depends on`.
Mainly that a downstream user cannot get rid of a `select`, but it's possible to re-define a symbol that has a `depends on` and thereby allow the setting to be available when alternative dependencies are met.

However I haven't looked into details regarding #48707 so above is based on general recommendations and experiences.

**Review by tejlmand — CHANGES_REQUESTED** [review]
currently blocking.

Although a discussion seems to have happened in #48707 then I was not consulted on such change and I need to take a closer look at this discussion because there are risks / pitfalls related to `select` when comparing to `depends on`.

I need to take a closer look at first #48707.

File: drivers/flash/Kconfig.at45
Comment (pdgendt) [reply drivers/flash/Kconfig.at45:10]: > Mainly that a downstream user cannot get rid of a `select`, but it's possible to re-define a symbol that has a `depends on` and thereby allow the setting to be available when alternative dependencies are met.

A downstream user can get rid of it, but only if the peripheral that selects the bus is `disabled` in the device tree or the driver `kconfig` symbol is manually set to `n`.

Comment (nordicjm) [reply drivers/flash/Kconfig.at45:10]: I had a check how this is done in linux, and it is done using `depends on`, here are some examples:
```
config TOUCHSCREEN_AD7879_I2C
        tristate "support I2C bus connection"
        depends on TOUCHSCREEN_AD7879 && I2C
        select REGMAP_I2C
        help
          Say Y here if you have AD7879-1/AD7889-1 hooked to an I2C bus.

          To compile this driver as a module, choose M here: the
          module will be called ad7879-i2c.

config TOUCHSCREEN_AD7879_SPI
        tristate "support SPI bus connection"
        depends on TOUCHSCREEN_AD7879 && SPI_MASTER
        select REGMAP_SPI
        help
          Say Y here if you have AD7879-1/AD7889-1 hooked to a SPI bus.

...

config MOUSE_PS2_SYNAPTICS_SMBUS
        bool "Synaptics PS/2 SMbus companion" if EXPERT
        default y
        depends on MOUSE_PS2 
        depends on I2C=y || I2C=MOUSE_PS2
        select MOUSE_PS2_SMBUS

...

config VIDEO_IMX219
        tristate "Sony IMX219 sensor support"
        depends on I2C && VIDEO_DEV
        select MEDIA_CONTROLLER
        select VIDEO_V4L2_SUBDEV_API
        select V4L2_FWNODE

...

menuconfig MEDIA_SUPPORT
        tristate "Multimedia support"
        depends on HAS_IOMEM

if MEDIA_SUPPORT

config MEDIA_SUBDRV_AUTOSELECT
        bool "Autoselect ancillary drivers (tuners, sensors, i2c, spi, frontends)"
        depends on HAS_IOMEM
        select I2C
        select I2C_MUX
        default y if MEDIA_SUPPORT_FILTER

...

config LTC2983
        tristate "Analog Devices Multi-Sensor Digital Temperature Measurement System"
        depends on SPI
        select REGMAP_SPI
```

**Comment by carlescufi** [comment]
Architecture WG:

- `depends on <BUS>` vs `select <bus>`
- The `<BUS>` is only enabled if the corresponding `DT_HAS_<>_ENABLED` Kconfig option is enabled, which it is only if there's a Devicetree node enabled
- This was done for all sensors in the tree, see [this commit](https://github.com/zephyrproject-rtos/zephyr/commit/df81fef94483da9f2811d48088affbcfd61ab18c)
- @henrikbrixandersen mentions that they use another compatible whenever they extend an upstream driver, which then decouples the OoT implementation from the original one
- @gmarull mentions how, before @galak's commit was merged, defconfig files were unmaintainable, and so going back to the old approach is not really maintainable
- @mbolivar-nordic mentions that there are workarounds for sensors with multiple buses: https://github.com/zephyrproject-rtos/zephyr/blob/3454a2547ead7f867c75cf83b12a38a6ea4dc599/drivers/sensor/bme280/Kconfig#L11

File: boards/arm/frdm_k82f/Kconfig.defconfig
Comment (danieldegrasse) [inline boards/arm/frdm_k82f/Kconfig.defconfig:23]: I'll leave the `select` vs `depends on` discussion to others, just wanted to make a note that the K82F changes here are valid given the other changes in this PR.

**Review by tejlmand — DISMISSED** [review]
A bit concerned on selecting higher level SPI from drivers.

One concern is the stuck-symbol which may occur when re-configuring.
Second concern is the `select` always kicking in, meaning users cannot get rid of a `select` as well as `select` overruling any `depends on` (a behavior which often surprises users)
There are some general design discussions / concerns as well, but would go to far to list everything here.

I acknowledge the limitation of current Kconfig / dts implementation wrt. scalability as described by https://github.com/zephyrproject-rtos/zephyr/commit/df81fef94483da9f2811d48088affbcfd61ab18c

To accommodate concerns stated above we should see if a corresponding hidden (promptless) symbol should be introduced in addition to, e.g `menuconfig SPI`.
But that is considered outside the scope of this PR.

Approving with the concerns given above.

**Review by carlescufi — DISMISSED** [review]

**Comment by carlescufi** [comment]
@pdgendt I've rebased the branch

**Review by henrikbrixandersen — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
Ref: https://github.com/zephyrproject-rtos/zephyr/pull/51598#pullrequestreview-1315442677

**Pull request merged**
